fdbserver/RocksDBLogForwarder.actor.cpp

```diff
@@ -0,0 +1,145 @@
+/*
+ * RocksDBLogForwarder.actor.cpp
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2013-2022 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "fdbserver/RocksDBLogForwarder.h"
+
+#include "flow/network.h"
+#include "flow/Trace.h"
+#include "fdbrpc/simulator.h"
+
+#include "flow/actorcompiler.h" // This must be the last include file
+
+using InfoLogLevel = rocksdb::InfoLogLevel;
+
+namespace {
+
+Severity getSeverityFromLogLevel(const InfoLogLevel& log_level) {
+	switch (log_level) {
+	case InfoLogLevel::DEBUG_LEVEL:
+		return SevDebug;
+	case InfoLogLevel::INFO_LEVEL:
+		return SevInfo;
+	case InfoLogLevel::WARN_LEVEL:
+		return SevWarn;
+	case InfoLogLevel::ERROR_LEVEL:
+		return SevError;
+	case InfoLogLevel::FATAL_LEVEL:
+		return SevError;
+	case InfoLogLevel::HEADER_LEVEL:
+		return SevVerbose;
+	case InfoLogLevel::NUM_INFO_LOG_LEVELS:
+		ASSERT(false);
+	}
+	UNREACHABLE();
+}
+
+} // namespace
+
+namespace details {
+
+void logTraceEvent(const RocksDBLogRecord& record) {
+	TraceEvent event = TraceEvent(record.severity, "RocksDBLogRecord", record.uid);
+	event.detail("RocksDBLogTime", record.logReceiveTime);
+
+	{
+		std::stringstream ss;
+		ss << record.threadID;
+		std::string threadID;
+		event.detail("RocksDBThreadID", threadID);
+	}
+
+	for (const auto& [k, v] : record.kvPairs) {
+		event.detail(k.c_str(), v);
+	}
+}
+
+ACTOR Future<Void> rocksDBPeriodicallyLogger(RocksDBLogger* pRecords) {
```

Comment:
If this is running in the main thread and consuming messages from a rocksdb thread, that definitely breaks determinism. I guess that's ok if this is rocksdb only which I think is already not deterministic.

If we're going this route I would personally recommend using a condition variable and notify instead of polling

Reply:
If RocksDB thread itself calls `deterministicRandom`, then the determinism is over; otherwise this code change should not impact since it only calls `deterministicRandom` in the main thread, when evaluating `TraceEvent` fields.